Source/JavaScriptCore/runtime/JSCell.cpp

```diff
@@ -338,4 +338,33 @@ NEVER_INLINE NO_RETURN_DUE_TO_CRASH NOT_TAIL_CALLED void reportZappedCellAndCras
 }
 #endif // CPU(X86_64)
 
+#if USE(JSVALUE32_64)
+bool isLiveConcurrently(const HeapCell* cell)
+{
+    if (!cell)
+        return false;
+
+    if (!isCompilationThread() || !Options::useConcurrentJIT())
+        return true;
+
+    // This also prevents pointer with invalid ranges, such as 0x02
+    MarkedBlock* candidate = MarkedBlock::blockFor(cell);
+    if (!candidate)
+        return false;
+
+    if (cell->isPreciseAllocation())
+        return cell->preciseAllocation().isLive();
+
+    Locker locker { candidate->lock() };
```

Comment:
This is not correct since there is no guarantee that candidate is a valid MarkedBlock pointer.

Reply:
We use `MarkedBlock::blockFor(cell)` to get the right candidate, wouldn't it prevent us from getting an invalid MarkedBlock pointer?

Reply:
It does not prevent us from getting an invalid MarkedBlock pointer.

Reply:
Alright, does it mean that I need to do something like `isPointerGCObjectJSCell`?
```
        if (pointer->isPreciseAllocation()) {
            auto* set = heap.objectSpace().preciseAllocationSet();
            ASSERT(set);
            if (set->isEmpty())
                return false;
            if (!set->isValidValue(pointer))
                return false;
            return set->contains(pointer);
        }
```
I tried that but the `set` I get is always `0x0`.

Reply:
Ditto, Alright, does it mean that I need to do something like `isPointerGCObjectJSCell`?
```
        MarkedBlock* candidate = MarkedBlock::blockFor(pointer);
        if (filter.ruleOut(bitwise_cast<uintptr_t>(candidate))) {
            ASSERT(!candidate || !set.contains(candidate));
            return false;
        }
```

Reply:
Here I added the filter to ruleOut bad candidates, like it's done in `isPointerGCObjectJSCell`